### PR TITLE
diffusers#7426 fix stable diffusion xl inference on MPS when dtypes shift unexpectedly due to pytorch bugs

### LINKED ISSUE #7426: [MPS] SDXL pipeline fails inference in fp16 mode
### Describe the bug

When running a pipeline under 🤗 Accelerate, my M3 system crashes with a tensor dtype mismatch during validation inference.

This issue does not occur without 🤗 Accelerate.

### Reproduction

Unfortunately, the way the reproducer is executed within a training script makes it exceedingly difficult/time-consuming to put together a small reproducer.

I added print statements to the pipeline so that I could determine where the broadcast compatibility error originates, and discovered this line:

```py
                print("compute the previous noisy sample x_t -> x_t-1")
                latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
```

which seems to bump the dtype from float16 to float32 on the 2nd step. the 1st step is correct and fine.

just on a whim, i modified the code to the following:

```py
                print("compute the previous noisy sample x_t -> x_t-1")
                old_dtype = latents.dtype
                latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
                if latents.dtype is not old_dtype:
                    latents = latents.to(old_dtype)
```

which then allows validations to complete and finish.

i'm sure there's a deeper issue and this is not a valid solution, so i have not opened up a PR with this solution.

it seems like the DDIM scheduler simply should not change the dtype.

### Logs

```shell
Before resolution:


2024-03-21 10:14:22,290 [DEBUG] (validation) Generating validation image: 
 Device allocations:
 -> unet on mps:0
 -> text_encoder on None
 -> vae on mps:0
 -> current_validation_prompt_embeds on mps:0
 -> current_validation_pooled_embeds on mps:0
 -> validation_negative_prompt_embeds on mps:0
 -> validation_negative_pooled_embeds on mps:0
2024-03-21 10:14:22,361 [DEBUG] (validation) Generating validation image: 
 Weight dtypes:
 -> unet: torch.float16
 -> text_encoder: None
 -> vae: torch.float16
 -> current_validation_prompt_embeds: torch.float16
 -> current_validation_pooled_embeds: torch.float16
 -> validation_negative_prompt_embeds: torch.float16
 -> validation_negative_pooled_embeds: torch.float16
2024-03-21 10:14:22,361 [DEBUG] (validation) Generating validation image: 
 -> Number of images: 1
 -> Number of inference steps: 30
 -> Guidance scale: 7.5
 -> Guidance rescale: 0.7
 -> Resolution: 1024.0
 -> Extra validation kwargs: {'generator': <torch._C.Generator object at 0x6cb88de10>}
1. Check inputs. Raise error if not correct
Prompt embeds shape: torch.Size([1, 77, 2048])
Negative prompt embeds shape: torch.Size([1, 77, 2048])
2. Define call parameters
3. Encode input prompt
4. Prepare timesteps
5. Prepare latent variables
6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
7. Prepare added time ids & embeddings
8. Denoising loop
8.1 Apply denoising_end
9. Optionally get Guidance Scale Embedding
expand the latents if we are doing classifier free guidance
predict the noise residual
 -> dtype: torch.float16
 -> dtype (latents): torch.float16
 -> dtype (prompt_embeds): torch.float16
 -> dtype (add_text_embeds): torch.float16
perform guidance
Based on 3.4. in https://arxiv.org/pdf/2305.08891.pdf
compute the previous noisy sample x_t -> x_t-1
call the callback, if provided
expand the latents if we are doing classifier free guidance
predict the noise residual
 -> dtype: torch.float32
 -> dtype (latents): torch.float32
 -> dtype (prompt_embeds): torch.float16
 -> dtype (add_text_embeds): torch.float16
loc("mps_add"("(mpsFileLoc): /AppleInternal/Library/BuildRoots/ce725a5f-c761-11ee-a4ec-b6ef2fd8d87b/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm":233:0)): error: input types 'tensor<2x1280xf32>' and 'tensor<1280xf16>' are not broadcast compatible
LLVM ERROR: Failed to infer result type(s).
```

After resolution:

```
 Weight dtypes:
 -> unet: torch.float16
 -> text_encoder: None
 -> vae: torch.float16
 -> current_validation_prompt_embeds: torch.float16
 -> current_validation_pooled_embeds: torch.float16
 -> validation_negative_prompt_embeds: torch.float16
 -> validation_negative_pooled_embeds: torch.float16
2024-03-21 10:12:17,921 [DEBUG] (validation) Generating validation image:
 -> Number of images: 1
 -> Number of inference steps: 30
 -> Guidance scale: 7.5
 -> Guidance rescale: 0.7
 -> Resolution: 1024.0
 -> Extra validation kwargs: {'generator': <torch._C.Generator object at 0x35ae4be50>}
1. Check inputs. Raise error if not correct
Prompt embeds shape: torch.Size([1, 77, 2048])
Negative prompt embeds shape: torch.Size([1, 77, 2048])
2. Define call parameters
3. Encode input prompt
4. Prepare timesteps
5. Prepare latent variables
6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
7. Prepare added time ids & embeddings
8. Denoising loop
8.1 Apply denoising_end
9. Optionally get Guidance Scale Embedding
expand the latents if we are doing classifier free guidance
predict the noise residual
 -> dtype: torch.float16
 -> dtype (latents): torch.float16
 -> dtype (prompt_embeds): torch.float16
 -> dtype (add_text_embeds): torch.float16
perform guidance
Based on 3.4. in https://arxiv.org/pdf/2305.08891.pdf
compute the previous noisy sample x_t -> x_t-1
call the callback, if provided
expand the latents if we are doing classifier free guidance
predict the noise residual
 -> dtype: torch.float16
 -> dtype (latents): torch.float16
 -> dtype (prompt_embeds): torch.float16
 -> dtype (add_text_embeds): torch.float16
```

### System Info

- `diffusers` version: 0.26.3
- Platform: macOS-14.4-arm64-arm-64bit
- Python version: 3.10.13
- PyTorch version (GPU?): 2.2.1 (False)
- Huggingface_hub version: 0.21.4
- Transformers version: 4.40.0.dev0
- Accelerate version: 0.26.1
- xFormers version: not installed
- Using GPU in script?: M3 Max
- Using distributed or parallel set-up in script?: False

### Who can help?

@pcuenca

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

Fixes #7426 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by bghira** [comment]
@sayakpaul i included the other pipelines just to keep them consistent even though normally inference does not hit this problem. it is likely nice to fix it for the edge cases where it would

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 1]
```diff
@@ -1193,7 +1193,11 @@ def __call__(
                     noise_pred = rescale_noise_cfg(noise_pred, noise_pred_text, guidance_rescale=self.guidance_rescale)
 
                 # compute the previous noisy sample x_t -> x_t-1
+                old_dtype = latents.dtype
                 latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
+                if latents.dtype != old_dtype:
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: Should this be guarded with `torch.backends.mps.is_available()` as well? It really seems to be happening only when `mps` is picked up no?

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 2]
```diff
@@ -1228,6 +1232,9 @@ def __call__(
             if needs_upcasting:
                 self.upcast_vae()
                 latents = latents.to(next(iter(self.vae.post_quant_conv.parameters())).dtype)
+            elif torch.backends.mps.is_available() and latents.dtype != self.vae.dtype:
+                # some platforms (eg. apple mps) misbehave due to a pytorch bug, this is a workaround
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1236]: Would be also nice to include a link if there's any public report of this bug on the PyTorch GitHub repository.

**Review by sayakpaul — APPROVED** [review]
To me the changes seem realistic and minimal enough to enable training on a different accelerator. So, I would be in favor of supporting this. 

@yiyixuxu @pcuenca WDYT?

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py (hunk 2 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1236]: https://github.com/pytorch/pytorch/pull/99272 is involved, not sure if there are additional issues.

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py (hunk 1 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: (nit: I'd maybe call it `latents_dtype`; `old` sounds almost like it'd be ok if it changes).

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: > Should this be guarded with `torch.backends.mps.is_available()`

In my opinion, it's ok the way it is as the comment already mentions `mps`, but no strong opinion.

**Review by pcuenca — APPROVED** [review]
Agree to add these workarounds to unblock use on mps. Thanks @bghira!

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py (hunk 1 above)
Comment (bghira) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: @sayakpaul @pcuenca it's about the outcome. if any other accelerator behaves in this way, would you rather it crash out of the box so that an issue is filed, or should we just invisibly fix it when we find it?

Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: How about throwing a warning so that the users are at least aware of it?

Comment (bghira) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py]: or raise a specific error asking them to file a report?

Comment (bghira) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py]: i modified the code so that it only executes on mps, as i figured it would be better to have full visibility into any other platform's dtype issues after further consideration.

the current error however, is pretty vague. it says two types are broadcast incompatible, which isn't very helpful to a new user

Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py]: Yeah raising an error sounds like a better idea. @pcuenca WDYT?

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py [hunk 3]
```diff
@@ -1193,7 +1193,11 @@ def __call__(
                     noise_pred = rescale_noise_cfg(noise_pred, noise_pred_text, guidance_rescale=self.guidance_rescale)
 
                 # compute the previous noisy sample x_t -> x_t-1
+                latents_dtype = latents.dtype
                 latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
+                if torch.backends.mps.is_available() and latents.dtype != latents_dtype:
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: So, if we were raise an error, I would do:

```suggestion
                if latents.dtype != latents_dtype:
                    if torch.backends.mps.is_available():
                        # some platforms (eg. apple mps) misbehave due to a pytorch bug: https://github.com/pytorch/pytorch/pull/99272
                        latents = latents.to(latents_dtype)
                    else:
                        raise ValueError("For the given accelerator, there seems to be 
an unexpected problem in type-casting. Please file an issue on the PyTorch GitHub repository. See also: https://github.com/huggingface/diffusers/pull/7446/.")
```

Does this make sense?

**Comment by sayakpaul** [comment]
@yiyixuxu could you also give this a look?

**Review by yiyixuxu — COMMENTED** [review]
thanks!

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py:1198]: @bghira a gentle ping here.

**Comment by sayakpaul** [comment]
@bghira okay if I applied the safe-guarding logic to the rest of the scripts and prep the PR for merging?

**Comment by bghira** [comment]
yup

**Comment by sayakpaul** [comment]
@bghira give this a final look and I will merge then?

**Comment by bghira** [comment]
nice, lgtm

**Comment by sayakpaul** [comment]
Thanks for your contributions, @bghira!
